underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** The output stream to a local file where the file will be buffered. */
+  private OutputStream mLocalOutputStream;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  private long mOffset;
+
+  /**
+   * Give each upload request an unique and continuous id
+   * so that S3 knows the part sequence to concatenate the parts to a single object.
+   */
+  private AtomicInteger mPartNumber;
+
+  /** The upload id of this multipart upload. */
+  private String mUploadId;
+
+  /**
+   * Constructs a new stream for writing a file.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param s3Client the Amazon S3 client to upload the file with
+   * @param executor a thread pool executor
+   */
+  public S3ALowLevelOutputStream(String bucketName, String key, AmazonS3 s3Client,
+      ExecutorService executor) {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mClient = s3Client;
+    mExecutor = MoreExecutors.listeningDecorator(executor);
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
+      mHash = null;
+    }
+    mKey = key;
+    // Partition size should be at least 5 MB, since S3 low-level multipart upload does not
+    // accept intermediate part smaller than 5 MB.
+    long partSize = Configuration.getBytes(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE);
+    mPartitionSize = partSize < 5 * Constants.MB ? 5 * Constants.MB : partSize;
+    mPartNumber = new AtomicInteger(1);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    mSingleCharWrite[0] = (byte) b;
+    write(mSingleCharWrite, 0, 1);
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (mUploadId == null) {
+      initMultiPartUpload();
+    }
+    if (b == null || len == 0) {
+      return;
+    }
+    if (mFile == null) {
+      initNewFile();
+    }
+    if (mOffset + len < mPartitionSize) {
+      mLocalOutputStream.write(b, off, len);
+      mOffset += len;
+    } else {
+      int firstLen = (int) (mPartitionSize - mOffset);
+      mLocalOutputStream.write(b, off, firstLen);
+      mOffset += firstLen;
+      uploadPart();
+      write(b, off + firstLen, len - firstLen);
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    mLocalOutputStream.flush();
```

Comment:
If `flush` is called immediately after `write`, then `mLocalOutputStream` is `null`.

Reply:
The mLocalOutputStream do have the chance to be null although it may be low.
Add a null check to make it more robust.
Changed to
```
    if (mLocalOutputStream != null) {
      mLocalOutputStream.flush();
    }
```
Thanks